File: php/routes/ClientLogin.php
<?php
/*
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2023, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE
*/

class ClientLogin {
  function init($f3){
    $f3->set('title', 'Nervatura Client custom token login');
    $f3->set('database', 'demo');
    $f3->set('username', 'admin');
    $f3->set('response_type', 'code');
    $f3->set('error', NULL);
    $f3->set('env', array(
      'NT_EXAMPLE_TOKEN_PRIVATE_KEY'=>getenv('NT_EXAMPLE_TOKEN_PRIVATE_KEY'),
      'NT_EXAMPLE_TOKEN_EXP'=>getenv('NT_EXAMPLE_TOKEN_EXP'),
      'NT_EXAMPLE_TOKEN_ALGORITHM'=>getenv('NT_EXAMPLE_TOKEN_ALGORITHM'),
      'NT_TOKEN_ISS'=>getenv('NT_TOKEN_ISS'),
      'NT_TOKEN_PUBLIC_KID'=>getenv('NT_TOKEN_PUBLIC_KID'),
      'NT_TOKEN_PUBLIC_KEY'=>getenv('NT_TOKEN_PUBLIC_KEY'),
      'NT_ALIAS_DEMO'=>getenv('NT_ALIAS_DEMO')
    ));
  }
  
  public function get($f3){
    $this->init($f3);
    $error = $f3->get('GET.error');
    if($error){
      $f3->set('error', base64_decode($error));
    }
    $this->sendResult($f3);
  }

  public function redirect($f3){
    $username = $f3->get('POST.username');
    $database = $f3->get('POST.database');
    $response_type = $f3->get('POST.response_type');
    $token = (new Utils())->CreateToken(array(
      'username' => $username, 
      'database' => $database,
      'algorithm' => getenv('NT_EXAMPLE_TOKEN_ALGORITHM'),
      'kid' => getenv('NT_TOKEN_PUBLIC_KID')
    ));
    $url = 'http://localhost:'.getenv('NT_HTTP_PORT').'/client';
    if ($response_type == 'code') {
      $code = uniqid();
      $cache = \Cache::instance();
      $cache->set('code', $code);
      $cache->set('token', $token);
      $url = $url.'?#code='.$code.'&callback=http://'.getenv('NT_EXAMPLE_HOST').':'.getenv('NT_EXAMPLE_PORT').'/client_login/token&error=http://'.getenv('NT_EXAMPLE_HOST').':'.getenv('NT_EXAMPLE_PORT').'/client_login';
    } else {
      $url = $url.'?#access_token='.$token.'&callback=http://'.getenv('NT_EXAMPLE_HOST').':'.getenv('NT_EXAMPLE_PORT').'/client_login';
    }
    
    $f3->reroute($url);
  }

  public function token($f3){
    $cache = \Cache::instance();
    $code = $cache->get('code');
    $body = json_decode(file_get_contents('php://input'), true);
    if(array_key_exists('code', $body) && ($body['code'] == $code)) {
      $token = $cache->get('token');
      $data = json_encode(array(
        'access_token' => $token,
        'token_type' => 'bearer',
        'expires_in' => getenv('NT_EXAMPLE_TOKEN_EXP'),
        'callback' => 'http://'.getenv('NT_EXAMPLE_HOST').':'.getenv('NT_EXAMPLE_PORT').'/client_login'
      ));
      header('Content-type: application/json', true, 200);
      echo $data;
      return;
    }
    header('Content-type: application/json', true, 400);
    $data = json_encode(array(
      'code' => 400,
      'error' => array(
        'message' => 'Invalid code'
      )
    ));
    echo $data;
  }

  function sendResult($f3){
    $f3->set('content', 'client_login.html');
    echo Template::instance()->render('layout.html');
  }
}

?>

File: go/app/client_login.go
package app

import (
	"encoding/base64"
	"net/http"
	"os"

	ut "github.com/nervatura/nervatura-examples/utils"
)

func (app *App) client_login(w http.ResponseWriter, r *http.Request) {
	data := map[string]interface{}{
		"username":      "admin",
		"database":      "demo",
		"response_type": "code",
		"title":         "Nervatura Client custom token login - with the HMAC algorithm",
		"error":         nil,
		"env": map[string]interface{}{
			"NT_EXAMPLE_TOKEN_PRIVATE_KEY": os.Getenv("NT_EXAMPLE_TOKEN_PRIVATE_KEY"),
			"NT_EXAMPLE_TOKEN_EXP":         os.Getenv("NT_EXAMPLE_TOKEN_EXP"),
			"NT_EXAMPLE_TOKEN_ALGORITHM":   os.Getenv("NT_EXAMPLE_TOKEN_ALGORITHM"),
			"NT_TOKEN_ISS":                 os.Getenv("NT_TOKEN_ISS"),
			"NT_TOKEN_PUBLIC_KID":          os.Getenv("NT_TOKEN_PUBLIC_KID"),
			"NT_TOKEN_PUBLIC_KEY":          os.Getenv("NT_TOKEN_PUBLIC_KEY"),
			"NT_ALIAS_DEMO":                os.Getenv("NT_ALIAS_DEMO"),
		},
	}
	if r.Method == "GET" {
		if r.URL.Query().Get("error") != "" {
			errMsg, err := base64.URLEncoding.DecodeString(r.URL.Query().Get("error"))
			if err == nil {
				data["error"] = string(errMsg)
			}
		}
		app.render(w, "client_login", data)
		return
	}

	err := r.ParseForm()
	if err != nil {
		data["error"] = err.Error()
		app.render(w, "client_login", data)
		return
	}
	for key := range r.PostForm {
		data[key] = r.PostForm.Get(key)
	}

	url := "http://localhost:" + os.Getenv("NT_HTTP_PORT") + "/client"
	//url := "http://localhost:3000"
	token, err := app.createToken(
		ut.ToString(data["username"], ""),
		ut.ToString(data["database"], ""),
		os.Getenv("NT_TOKEN_PUBLIC_KID"),
		os.Getenv("NT_EXAMPLE_TOKEN_ALGORITHM"))
	if err != nil {
		data["error"] = err.Error()
		app.render(w, "client_login", data)
		return
	}
	if ut.ToString(data["response_type"], "") == "code" {
		code := ut.RandString(10)
		app.sessions[code] = token
		url += "?#code=" + code + "&callback=http://" + os.Getenv("NT_EXAMPLE_HOST") + ":" + os.Getenv("NT_EXAMPLE_PORT") + "/client_login/token&error=http://" + os.Getenv("NT_EXAMPLE_HOST") + ":" + os.Getenv("NT_EXAMPLE_PORT") + "/client_login"
	} else {
		url += "?#access_token=" + token + "&callback=http://" + os.Getenv("NT_EXAMPLE_HOST") + ":" + os.Getenv("NT_EXAMPLE_PORT") + "/client_login"
	}
	http.Redirect(w, r, url, http.StatusSeeOther)
}

func (app *App) codeToken(w http.ResponseWriter, r *http.Request) {
	var data map[string]interface{}
	err := ut.ConvertFromReader(r.Body, &data)
	if err != nil {
		app.sendRequest(w, 400, map[string]interface{}{"code": 400, "error": map[string]string{"message": err.Error()}})
		return
	}
	if code, found := data["code"]; found {
		if token, found := app.sessions[ut.ToString(code, "")]; found {
			app.sendRequest(w, 200, map[string]interface{}{
				"access_token": token,
				"token_type":   "bearer",
				"expires_in":   os.Getenv("NT_EXAMPLE_TOKEN_EXP"),
				"callback":     "http://" + os.Getenv("NT_EXAMPLE_HOST") + ":" + os.Getenv("NT_EXAMPLE_PORT") + "/client_login",
			})
			return
		}
	}
	app.sendRequest(w, 400, map[string]interface{}{"code": 400, "error": map[string]string{"message": "Invalid code"}})
}
